src/Raven.Server/Documents/Queries/QueriedDocument.cs

```diff
@@ -0,0 +1,133 @@
+    private QueriedDocument _referencingSingle;
+    private List<QueriedDocument> _referencing;
+
+    public bool CanDispose => _refCount == 1;
+    public bool IsCloned;
+
+    public void LinkReferencedDocument(QueriedDocument additionalDocument)
+    {
+        // loaded document doesn't exists
+        if (additionalDocument is null)
+            return;
+
+        // Since most of queries loads only one document
+        // e.g. load originalDoc.ReferenceId as ref
+        // We can avoid creating a list object for single reference and store it in variable.
+        if (_referencingSingle == null)
+        {
+            _referencingSingle = additionalDocument;
+        }
+        else if (ReferenceEquals(_referencingSingle, ExplicitNullReference) == false)
+        {
+            _referencing ??= new();
+            _referencing.Add(_referencingSingle);
+            // ExplicitNullReference means we've multiple references.
+            _referencingSingle = ExplicitNullReference;
+        }
+        else
+        {
+            _referencing.Add(additionalDocument);
+        }
+
+        additionalDocument.IncreaseReference();
+    }
+
+    public override TDocument Clone<TDocument>(JsonOperationContext context)
+    {
+        var data = Data.Clone(context);
+        return CloneWith<TDocument>(context, data);
+    }
+
+
+    public override TDocument CloneWith<TDocument>(JsonOperationContext context, BlittableJsonReaderObject newData)
+    {
+        if (typeof(TDocument) != typeof(QueriedDocument))
+            return base.Clone<TDocument>(context);
+
+        IncreaseReference();
+
+        var cloned = (QueriedDocument)(object)base.CloneWith<TDocument>(context, newData);
+
+        cloned.IsCloned = true;
+        cloned._sourceDocument = this;
+
+        //We cloned document for projection purposes. That indicates that all loaded references are related to the copy.
+        cloned._referencing = _referencing;
+        _referencing = null;
+        cloned._referencingSingle = _referencingSingle;
+        _referencingSingle = null;
+
+        return (TDocument)(object)cloned;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public void IncreaseReference() => _refCount++;
+
+    public override void Dispose()
+    {
+        _refCount--;
+        Debug.Assert(_refCount >= 0, "This document is already disposed, however we kept reference in other document. This should not be possible.");
+
+        if (_sourceDocument != null)
+        {
+            Debug.Assert(IsCloned);
+            Debug.Assert(_refCount == 0, "We should not have reference copy to clone.");
+            //Decrease reference at parent document.
+            Debug.Assert(_sourceDocument._refCount >= 2);
+            Debug.Assert(_sourceDocument.Data != null);
+
+            Debug.Assert(ReferenceEquals(_sourceDocument.Data, Data) == false);
+            _sourceDocument.Dispose();
+            Debug.Assert(_sourceDocument.Data != null);
```

Comment:
If any of those assertions fails one day during the tests you will not know which one. Please add messages as second parameter